### PR TITLE
On OnePlus phone, Android 4.3, there is a bug

### PR DESCRIPTION
**Pull request by icexile**
On OnePlus phone, Android 4.3, there is a bug that connect() returns -1 but errno == 0, and actually the tcp three-way handshake is completed.

### CODE REVIEW

File: src/unix/tcp.c [hunk 1]
```diff
@@ -162,7 +162,7 @@ int uv__tcp_connect(uv_connect_t* req,
     r = connect(uv__stream_fd(handle), addr, addrlen);
   while (r == -1 && errno == EINTR);
 
-  if (r == -1) {
+  if (r == -1 && errno) {
```
Comment (bnoordhuis) [inline src/unix/tcp.c]: You should `errno = 0;` before the call to connect() inside the do/while block.  Stylistically, can you check `errno != 0` here?

**Comment by bnoordhuis** [comment]
Can you add an explaining comment in the source?  Also, can you update the commit log to conform to our house style?

Does our test suite catch this bug?  If not, can you look into adding a test that does?  I.e., a test that fails without your patch and passes with.

**Comment by icexile** [comment]
OK, I will explain why we should do this check in the source code, but I can not add the test that fails without my patch, because this situation is only occured on OnePlus phone, version A0001_12_150227, which Android version is 4.3.

**Comment by icexile** [comment]
On other platforms, errno will not be zero when connect returns -1.

File: src/unix/tcp.c [hunk 2]
```diff
@@ -161,8 +161,12 @@ int uv__tcp_connect(uv_connect_t* req,
   do
     r = connect(uv__stream_fd(handle), addr, addrlen);
   while (r == -1 && errno == EINTR);
-
-  if (r == -1) {
+
```
Comment (bnoordhuis) [inline src/unix/tcp.c]: Remember, you need to set errno to zero before calling connect().

Comment (icexile) [inline src/unix/tcp.c]: OK

**Comment by bnoordhuis** [comment]
LGTM.  CI: https://ci.nodejs.org/job/libuv-test-commit/68/

Can you squash and update the commit log?  Thanks.

**Comment by iWuzHere** [comment]
CI looks happy to me.
LGTM once comments above are addressed (squashing and updating commit log)

**Comment by saghul** [comment]
LGTM (once the nits are addressed)

**Comment by icexile** [comment]
The commit log is update ~

**Comment by iWuzHere** [comment]
The first commit line is too long. Should be <=50 chars
See https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md under "COMMIT"

**Comment by icexile** [comment]
ok, fixed.

**Comment by saghul** [comment]
Since the code runs for all Unix versions, not Android only, maybe we should have "unix" as the subsystem?

**Comment by icexile** [comment]
Yes ~~ you are right, I have changed the commit log ~

**Comment by saghul** [comment]
LGTM

**Comment by iWuzHere** [comment]
LGTM

**Comment by saghul** [comment]
Landed in 3a4f2d3, thanks!

**Pull request closed without merging**